django-stubs/db/migrations/operations/models.pyi

```diff
@@ -1,58 +1,86 @@
+    @property
     def name_lower(self) -> str: ...
 
 class CreateModel(ModelOperation):
-    fields: Sequence[Tuple[str, Field]] = ...
-    options: Any = ...
-    bases: Optional[Sequence[Union[type, str]]] = ...
+    fields: List[Tuple[str, Field]] = ...
+    options: Dict[str, Any] = ...
+    bases: Optional[Sequence[Union[Type[Model], str]]] = ...
     managers: Optional[Sequence[Tuple[str, Manager]]] = ...
     def __init__(
         self,
         name: str,
-        fields: Sequence[Tuple[str, Field]],
+        fields: List[Tuple[str, Field]],
         options: Optional[Dict[str, Any]] = ...,
-        bases: Optional[Sequence[Union[type, str]]] = ...,
+        bases: Optional[Sequence[Union[Type[Model], str]]] = ...,
         managers: Optional[Sequence[Tuple[str, Manager]]] = ...,
     ) -> None: ...
-    def model_to_key(self, model: str) -> List[str]: ...
 
 class DeleteModel(ModelOperation): ...
 
 class RenameModel(ModelOperation):
-    old_name: Any = ...
-    new_name: Any = ...
+    old_name: str = ...
+    new_name: str = ...
     def __init__(self, old_name: str, new_name: str) -> None: ...
+    @property
     def old_name_lower(self) -> str: ...
+    @property
     def new_name_lower(self) -> str: ...
 
-class AlterModelTable(ModelOperation):
+class ModelOptionOperation(ModelOperation): ...
+
+class AlterModelTable(ModelOptionOperation):
     table: Optional[str] = ...
     def __init__(self, name: str, table: Optional[str]) -> None: ...
 
-class ModelOptionOperation(ModelOperation): ...
-class FieldRelatedOptionOperation(ModelOptionOperation): ...
+class AlterTogetherOptionOperation(ModelOptionOperation):
+    option_name: str = ...
+    def __init__(
+        self,
+        name: str,
+        option_value: _ListOrTuple[Tuple[str, str]],
+    ) -> None: ...
+    @property
+    def option_value(self) -> Set[Tuple[str, str]]: ...
+    def deconstruct(self) -> Tuple[str, Sequence[Any], Dict[str, Any]]: ...
+    def state_forwards(self, app_label: str, state: Any) -> None: ...
+    def database_forwards(
+        self, app_label: str, schema_editor: BaseDatabaseSchemaEditor, from_state: Any, to_state: Any
+    ) -> None: ...
+    def database_backwards(
+        self, app_label: str, schema_editor: BaseDatabaseSchemaEditor, from_state: Any, to_state: Any
+    ) -> None: ...
+    def references_field(self, model_name: str, name: str, app_label: str) -> bool: ...
+    def describe(self) -> str: ...
+    @property
+    def migration_name_fragment(self) -> str: ...
 
-class AlterUniqueTogether(FieldRelatedOptionOperation):
+class AlterUniqueTogether(AlterTogetherOptionOperation):
     option_name: str = ...
-    unique_together: Collection[Sequence[str]] = ...
-    def __init__(self, name: str, unique_together: Optional[Collection[Sequence[str]]]) -> None: ...
+    unique_together: _ListOrTuple[Tuple[str, str]] = ...
+    def __init__(self, name: str, unique_together: _ListOrTuple[Tuple[str, str]]) -> None: ...
```

Comment:
This isn't correct, Django 4.0 generates migrations using the `set` type, e.g.

```python
        migrations.AlterUniqueTogether(
            name='rewriterule',
            unique_together={('field1', 'field2')},
        ),
```

Causes mypy error:

> Argument "unique_together" to "AlterUniqueTogether" has incompatible type "Set[Tuple[str, str, str]]"; expected "Union[List[Tuple[str, str]], Tuple[Tuple[str, str], ...], Tuple[]]" [arg-type]

Same issue affects `AlterIndexTogether`.

Reply:
Ough. I have `migrations` excluded from typechecking so haven't noticed this issue. Thank you for reporting! Will fix soon. I have not respected `except TypeError: return option_together` in `models.options.normalize_together()`.